finat/fiat_elements.py

```diff
@@ -142,6 +144,98 @@ def point_evaluation(self, order, refcoords, entity=None):
+                        weight_array[key] = item
+                    weight_tensor = gem.Literal(weight_array)
+
+                pts_in_derivs.append((point_set, weight_tensor, alpha_tensor))
+            derivs.append(pts_in_derivs)
+
+            # For derivatives
+            deriv_dict_items = sorted(dual.deriv_dict.items())  # Ensure parallel safety
+            for derivative_order in range(1, max_deriv_order+1):
+                pts_in_derivs = []
+                # TODO: Combine tensors for tups of same derivative order
+                for pt, tups in deriv_dict_items:
+                    weights, alphas, cmps = zip(*tups)
+
+                    # TODO: Repeated points and repeated tups
+                    # Get evaluations of this derivative order
+                    weight, alpha, cmp = [], [], []
+                    for j, a in enumerate(alphas):
+                        if sum(a) == derivative_order:
+                            weight.append(weights[j])
+                            alpha.append(alphas[j])
+                            cmp.append(cmps[j])
+                    if len(alpha) == 0:
+                        continue
+
+                    # alpha_tensor assumes weights for each derivative component are equal
+                    # TODO: Case for unequal weights
+                    if len(alpha) > 1:
+                        if 0.0 not in weight:
+                            assert np.isclose(min(weight), max(weight))
+                        else:
+                            non_zero_weight = [w for w in weight if w != 0.0]
+                            assert np.isclose(min(non_zero_weight), max(non_zero_weight))
+
+                    # For direct indexing
+                    alpha_idx = tuple(tuple(chain(*[(j,)*a for j, a in enumerate(alph)])) for alph in alpha)
+                    # TODO: How to ensure correct combination (indexing) with weight_tensor?
+                    alpha_arr = np.zeros((self._element.ref_el.get_spatial_dimension(),) * derivative_order)
+                    for idx in alpha_idx:
+                        alpha_arr[idx] = 1
+                    alpha_tensor = gem.Literal(alpha_arr)
+
+                    try:
+                        point_set = point_set_cache[(pt,)]
+                    except KeyError:
+                        point_set = PointSet((pt,))
+                        point_set_cache[(pt,)] = point_set
+
+                    weight_dict = {c: w for w, c in zip(weight, cmp)}
+                    # Each entry of tensor is weight of that component
+                    if len(self.value_shape) == 0:
+                        weight_tensor = gem.Literal(weight_dict[tuple()])
+                    else:
+                        weight_array = np.zeros(self.value_shape)
+                        for key, item in weight_dict.items():
+                            weight_array[key] = item
+                        weight_tensor = gem.Literal(weight_array)
```

Comment:
Longwinded way of saying `gem.Literal(np.asarray(weight))`?

Reply:
I think so; the dual basis for FIAT elements with derivatives is a bit muddy, it'll need some restructuring